File: backend/app/api/notes.py
"""``/api/v1/notes`` — the five routes, and nothing else.

Every decision with teeth in it lives one module away: identity in ``refs.py``, authorization in
``app/auth/authorization.py``, the error shape in ``errors.py``, the payload in ``schemas.py``. What
is left here is genuinely just wiring, which is the point — a route that is four lines long has
nowhere to hide a fifth spelling of a rule.

Four things to notice, because all four are load-bearing:

- **No route parses an identifier.** ``NoteFromRef`` resolves it, checks it and hands back a
  ``Note``. That is ADR 0008's "resolve centrally, not per call site" made unavoidable rather than
  merely recommended: there is no identifier in scope to get wrong.
- **No route builds a note query.** A list composes onto ``notes_owned_by``, which already carries
  ``WHERE owner_id = :caller``; ``tests/unit/test_no_unscoped_note_query.py`` fails the build if
  this file ever names ``Note`` inside a ``select()``. Search (KAN-558) is ``notes_matching``, which
  is that same statement with two clauses added, so it is one ``WHERE`` and not two.
- **No route decides a conflict.** ADR 0009's precondition is one call to ``enforce_precondition``,
  so a second write endpoint gets the guarantee by making the same call rather than by
  reimplementing a comparison.
- **No route decides what an empty ``?q=`` means.** ``app/api/search.py`` does, before this module
  is reached, and hands the list route ``None`` or a term that is known to be non-blank.

`/links` and `/backlinks` landed in KAN-566 and are **not** in this file: see ``app/api/links.py``,
which argues the split. The short version is that they are the only routes under ``app/api/`` that
take a bearer and an upstream client as well as a session, so the three-phase body that follows from
that reads as what it is in its own module and as an exception here. They did depend on
``NoteFromRef`` and inherit ADR 0008 for free, exactly as this docstring predicted.

``/versions`` landed in KAN-1064 and **is** in this file, unlike ``/links``/``/backlinks`` — it
needs no bearer and no upstream client, just a scoped read over ``note_version``
(``app/note_versions.py``), so it is the plain wiring this module's docstring describes rather than
the exception ``links.py`` is.

Deliberately absent: paging of any shape.
"""

from typing import Annotated

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.orm import Session

from app.api.concurrency import enforce_precondition
from app.api.refs import NoteFromRef
from app.api.schemas import (
    NoteCreate,
    NoteList,
    NoteRead,
    NoteUpdate,
    NoteVersionList,
    NoteVersionRead,
)
from app.api.search import SearchTerm
from app.auth import (
    Principal,
    TeamAccessResolver,
    get_principal,
    get_team_access_resolver,
    notes_matching,
    notes_owned_by,
)
from app.db import get_session
from app.integrations.dependencies import CallerBearer
from app.models import Note
from app.note_links import reconcile_note_links, resolve_pending_note_links
from app.note_versions import cut_version, note_versions

router = APIRouter(prefix="/api/v1", tags=["notes"])

CurrentPrincipal = Annotated[Principal, Depends(get_principal)]
DbSession = Annotated[Session, Depends(get_session)]
CurrentTeamResolver = Annotated[TeamAccessResolver, Depends(get_team_access_resolver)]


@router.post(
    "/notes",
    status_code=status.HTTP_201_CREATED,
    summary="Create a note",
)
def create_note(
    payload: NoteCreate,
    principal: CurrentPrincipal,
    session: DbSession,
    response: Response,
) -> NoteRead:
    """Create a note owned by the caller.

    The ``ref``, the ``id`` and both timestamps are allocated by Postgres inside the INSERT and read
    back afterwards, never assembled here — ADR 0008 §Decision, and the reason two concurrent
    writers cannot be handed the same ref.

    The owner is the resolved principal and is not a request field. There is no route by which a
    caller can file a note against somebody else's UUID.

    KAN-562: the body's ``[[KAN-n]]`` / ``[[EPIC-n]]`` wikilinks are recorded in ``note_link`` in
    the same transaction as the note itself, via ``reconcile_note_links``. The explicit ``flush()``
    before that call is load-bearing — it is what makes ``note.id`` (the edges' foreign key)
    available before the transaction commits, rather than relying on the implicit flush inside
    ``commit()`` to have happened first.

    KAN-563: the same call also resolves any of *this* note's own ``[[Some Title]]`` links against
    notes that already exist, and ``resolve_pending_note_links`` afterward points any *other* note's
    still-unresolved link at this one, in case this note's title is exactly what one was waiting
    for. Both stay inside this same transaction for the reason above — a resolution landing without
    the note it names, or the reverse, is a state nothing here may produce.

    R13/KAN-1064: ``cut_version`` snapshots the created body as the note's first ``note_version``
    row, in the same transaction, so a note's history starts at creation rather than at its first
    edit. Every note gets one, even ``NoteCreate``'s default empty body — "on every body write, no
    heuristic" (BREADBOARD.md's R13) draws no exception for a body that happens to be ``""``.
    """
    note = Note(owner_id=principal.id, **payload.model_dump())
    session.add(note)
    session.flush()
    reconcile_note_links(session, note)
    resolve_pending_note_links(session, note)
    cut_version(session, note)
    session.commit()
    session.refresh(note)

    # The canonical form, which the ref resolver accepts back verbatim (ADR 0008's round trip).
    response.headers["Location"] = f"{router.prefix}/notes/{note.ref}"
    return NoteRead.of(note)


@router.get("/notes", summary="List the caller's notes, or search them with ?q=")
def list_notes(
    principal: CurrentPrincipal,
    session: DbSession,
    term: SearchTerm,
    bearer: CallerBearer,
    team_resolver: CurrentTeamResolver,
) -> NoteList:
    """Every note the caller owns, newest first — or, with ``?q=``, the ones that match it.

    Scoped in SQL by ``notes_owned_by``, not filtered afterwards: SLICES §V1 asks for another user's
    note to be *omitted* rather than fetched and hidden, and only the ``WHERE`` can say which
    happened. Composing here cannot lose that clause (KAN-535), and ``notes_matching`` composes onto
    the very same statement, so KAN-558's "another user's matching note must never appear" is that
    one clause rather than a second implementation of it.

    **Team membership is resolved before this session has run a single query** (ADR 0011, R16.3) —
    unlike the single-note path in ``app/api/refs.py``, there is no owner check to fail first and
    defer on, since a list has to know every team it should widen for before it can build the
    ``WHERE``. No connection is held across the call for the same reason it never is in
    ``resolve_note``: nothing here has queried yet, so there is nothing to release.

    **Two orders, because they answer two different questions, and each one is deterministic.**

    - No ``q``: ``updated_at DESC, id DESC``. The second column is not decoration — two notes
      written inside one transaction share an ``updated_at`` (``now()`` is transaction start time,
      per the model's own comment), so without a tie-break the order would be whatever Postgres felt
      like, and V4's "identical queries return results in a deterministic order" would already be
      false here. ``KayaClient.list_notes``' docstring depends on this order.
    - With ``q``: ``ts_rank DESC, id DESC``. Relevance is the only useful order for a search and a
      meaningless one for a list, so this is one order per request rather than one rule bent twice.
      The tie-break is the same column for the same reason; see ``notes_matching``, which owns both
      halves of the search order so they cannot be applied separately.

    The branch below chooses a **statement** and stops — the shape ``app/api/refs.py``'s
    ``resolve_note`` uses, for the same reason: one ``session.scalars``, one ``NoteList``, so there
    is no second code path in which projection, scoping or the envelope could differ. What ``?q=``
    means when it is empty is decided in ``app/api/search.py`` and has already happened by the time
    this function runs.

    No paging parameter: no card has asked for one, and ``next_cursor`` is additive to this envelope
    when one does. It is deliberately not added *with* search either — a `limit` would need a
    documented interaction with ranking, and that is a second undiscussed contract.
    """
    team_ids = team_resolver.member_of(bearer) if bearer is not None else frozenset()
    statement = (
        notes_owned_by(principal, team_ids).order_by(Note.updated_at.desc(), Note.id.desc())
        if term is None
        else notes_matching(principal, term, team_ids)
    )
    return NoteList(notes=[NoteRead.of(note) for note in session.scalars(statement)])


@router.get("/notes/{ref}", summary="Read one note by NOTE-n or by id")
def get_note(note: NoteFromRef) -> NoteRead:
    """One note, addressed as ``NOTE-12``, ``note-12`` or ``12``.

    The whole route is the dependency. Both spellings produce byte-identical bodies on a hit and
    byte-identical `404`s on a miss because there is nothing in this function that could tell them
    apart — see ``app/api/refs.py``.
    """
    return NoteRead.of(note)


@router.get(
    "/notes/{ref}/versions",
    summary="Every version of a note's body, newest first",
)
def list_note_versions(note: NoteFromRef, session: DbSession) -> NoteVersionList:
    """R13/KAN-1064: every snapshot ``cut_version`` has ever cut for this note.

    A four-line route over ``app/note_versions.py``'s ``note_versions``, the same shape every other
    route in this file is — the query, the scoping and the order are all decided one module away.
    ``note.id`` has already been through ``NoteFromRef`` and therefore ``authorize_note`` by the
    time it reaches ``note_versions``, which is what lets that function build a statement over a
    table with no owner column of its own and still be correctly scoped (see its docstring, and
    ``app/models/note_version.py``'s).

    Full bodies, not snippets: see ``NoteVersionRead``'s docstring for the preview-endpoint design
    call this response shape *is* — there is no separate preview route, on purpose.

    A note with exactly one save (its own creation) still returns one row, never an empty list —
    ``create_note`` cuts a version too, so "no history yet" is not a state this note can be in once
    it exists.
    """
    statement = note_versions(note.id)
    return NoteVersionList(
        versions=[NoteVersionRead.of(version) for version in session.scalars(statement)]
    )


@router.patch("/notes/{ref}", summary="Edit a note, or move it")
def update_note(note: NoteFromRef, payload: NoteUpdate, session: DbSession) -> NoteRead:
    """Change ``title``, ``body`` and/or ``path``. Omitted fields are left alone.

    Moving a note between folders is this route with ``{"path": "…"}`` and nothing else — one
    column, no link rewriting, nothing to break (ADR 0008). There is no separate move endpoint
    because there is no separate operation.

    **Two write semantics, and which one you get is the caller's choice** (ADR 0009). Send
    ``if_updated_at`` and the write is guarded: an ``updated_at`` that has moved on is a `409`
    carrying both versions, and nothing is written. Omit it and the write is a plain
    last-write-wins overwrite — specified, so that `curl` and a `kaya note edit` that omits
    `--if-updated-at` both work without a read-first dance. The precondition is a guarantee
    available to clients that want it, not a tax on every caller, so do not make it required.

    The guard runs **before** anything is applied, which is what makes a refused write atomic: a
    `PATCH` carrying a title and a body is rejected whole rather than leaving the title applied and
    the body not. See ``app/api/concurrency.py`` for why a metadata-only write is unguarded even
    when it carries a stale precondition.

    KAN-562: a write that touches ``body`` reconciles ``note_link`` in the same transaction —
    removed wikilinks disappear, added ones appear, and one left untouched is not re-written. A
    title- or path-only edit does not scan anything: ``find_wikilinks`` / ``find_note_title_links``
    read the body, and a save that never changed it cannot have changed what the body links to.

    KAN-563: a write that touches ``title`` calls ``resolve_pending_note_links`` — a rename can
    make this note satisfy some *other* note's still-unresolved ``[[Some Title]]`` link, and that
    other row would otherwise wait forever for a save of its own that has no reason to happen. This
    runs independently of whether ``body`` also changed in the same request, and it runs after
    ``note.title`` has already been assigned above, so it always resolves against the new value.

    R13/KAN-1064/1066: a write that touches ``body`` also calls ``cut_version`` — a snapshot of the
    body this write just set, in the same transaction as the update it belongs to. A title- or
    path-only edit cuts nothing, matching the wikilink reconcile immediately above it: neither one
    has a reason to run over a body that did not change. **This is also how a restore works** —
    KAN-1066's "restore a version" is nothing more than this same route called with ``body`` set to
    an old version's content, so a restore is guarded by the identical precondition, reconciles
    wikilinks the identical way, and cuts its own new version the identical way. No branch in this
    function knows or needs to know that a particular call was a restore.
    """
    enforce_precondition(session, note, payload)

    changes = payload.changes()
    for field, value in changes.items():
        setattr(note, field, value)

    # An empty PATCH is a legal no-op and must not restamp `updated_at` — that value is ADR 0009's
    # concurrency token, and a write that moves it without changing anything would invalidate every
    # other client's precondition for no reason. SQLAlchemy would emit no UPDATE anyway; committing
    # only when something changed says so out loud.
    if changes:
        if "body" in changes:
            reconcile_note_links(session, note)
            cut_version(session, note)
        if "title" in changes:
            resolve_pending_note_links(session, note)
        session.commit()
        session.refresh(note)

    return NoteRead.of(note)


@router.delete(
    "/notes/{ref}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_class=Response,
    summary="Delete a note",
)
def delete_note(note: NoteFromRef, session: DbSession) -> Response:
    """Delete the note. `204`, no body — there is nothing left to describe.

    The ref is **not** returned to the sequence. Refs are never reused (ADR 0008), so a later
    ``GET NOTE-12`` for a deleted note is a `404` forever rather than somebody else's note.
    """
    session.delete(note)
    session.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/lib/notes.ts
/**
 * The five note calls, one function each, and nothing else.
 *
 * This module holds no state and makes no decisions about what a caller sees — it is the typed
 * surface of `/api/v1/notes` and the place a component looks instead of assembling a path. Every
 * function returns a **complete record** (see `api.ts` on ADR 0004); none of them filters fields,
 * cuts prose or counts anything.
 */

import { apiRequest, type RequestOptions } from './api'
import type {
  Link,
  LinkList,
  Note,
  NoteCreate,
  NoteList,
  NoteUpdate,
  NoteVersion,
  NoteVersionList,
} from './types'

type Options = Pick<RequestOptions, 'signal' | 'fetchImpl'>

/**
 * The path for one note, with the ref as a **single** percent-encoded segment.
 *
 * `encodeURIComponent` rather than `encodeURI`, and shared by every ref-taking call for the same
 * reason `kaya-client` shares `_note_path`: a ref is one segment, so a `/` inside a bad ref must
 * become `%2F` and address a 404 rather than silently becoming a different route. `#NOTE-12` is a
 * documented `400` from the backend (ADR 0008), and it only gets there if the `#` survives the trip
 * — unencoded it would be a fragment the request never carries.
 */
export function notePath(ref: string): string {
  return `notes/${encodeURIComponent(ref)}`
}

/**
 * Every note the caller owns, newest first — or, with `q`, the ones that match it (KAN-559).
 *
 * `q` is forwarded to `GET /api/v1/notes?q=` exactly as it arrived: `undefined` adds no query
 * parameter at all, which is the same request this function always made, and a defined value is
 * sent verbatim, blank or not. `app/api/search.py` is the one place that decides what a blank `q`
 * means — a present-but-empty term is a `400 empty_search_query` — and this module has no second
 * opinion to disagree with it. **Whether to send `q` at all for an empty search box is one branch
 * the caller makes**, not this function: a cleared box must send no `q`, never `q=`.
 *
 * The response is the same `NoteList` shape a plain list gets, so this stays one function rather
 * than a second one for the search case.
 */
export async function listNotes(options: Options & { q?: string } = {}): Promise<Note[]> {
  const { q, ...rest } = options
  const path = q === undefined ? 'notes' : `notes?q=${encodeURIComponent(q)}`
  const payload = await apiRequest<NoteList>(path, rest)
  return payload.notes
}

/** One note, addressed as `NOTE-12`, `note-12` or `12` — all three resolve identically. */
export function getNote(ref: string, options: Options = {}): Promise<Note> {
  return apiRequest<Note>(notePath(ref), options)
}

export function createNote(input: NoteCreate, options: Options = {}): Promise<Note> {
  return apiRequest<Note>('notes', { ...options, method: 'POST', body: input })
}

/**
 * Change `title`, `body` and/or `path`. Omitted fields are left alone.
 *
 * Pass `if_updated_at` and the write is guarded: a stale value is a `409` carrying both versions
 * (`ApiError.details`), and nothing is written. Omit it and the write is a plain overwrite, **by
 * specification** — so there is no `force` parameter here, exactly as `kaya note edit` has no
 * `--force`. The unguarded write is spelled by not passing something.
 */
export function updateNote(ref: string, patch: NoteUpdate, options: Options = {}): Promise<Note> {
  return apiRequest<Note>(notePath(ref), { ...options, method: 'PATCH', body: patch })
}

/**
 * Move a note between folders. Sugar over {@link updateNote}, and it must stay sugar.
 *
 * ADR 0008: moving a note *is* a `PATCH` to `path`, with no link rewriting and no move endpoint.
 * It takes no precondition because ADR 0009 guards only writes touching `body`, so one here would
 * be accepted and ignored — a parameter that silently does nothing is worse than one that does not
 * exist.
 */
export function moveNote(ref: string, path: string, options: Options = {}): Promise<Note> {
  return updateNote(ref, { path }, options)
}

/**
 * Every note of the caller's whose body links to this one — `GET /notes/{ref}/backlinks` (KAN-568).
 *
 * **It returns `Note[]` because the API returns the same `NoteList` a plain list does**, so this is
 * `listNotes` at a different URL and there is no second type, no second envelope and no `Backlink`
 * interface anywhere in `frontend/`. `backend/app/api/links.py` argues that end of it: a backlink
 * *is* a note, and a link-shaped record would publish a second spelling of something the caller can
 * already read. `kaya-client` reached the same conclusion from the other side — its `backlinks()`
 * returns the note noun and the note columns, so `--fields` and `--full` arrived with no code
 * written for them.
 *
 * Two properties worth knowing before building on it, both `app.auth.notes_linking_to`'s rather than
 * this function's:
 *
 * - **It makes no upstream call.** "Which notes mention this one" is a join over two of kaya's own
 *   tables, so this request is answerable with pandan stopped and a cold cache (ADR 0003, SLICES
 *   §V5's R5.1). Every other read in this module needs pandan for authentication and this one does
 *   too — what it does not need is pandan for its *content*, which `/links` does.
 * - **The match key is `resolved_id`, never the title**, so renaming a note does not break the
 *   backlinks to it, and an edge whose `resolved_id` is still `NULL` is deliberately not a backlink
 *   to anything.
 *
 * Order is `updated_at DESC, id DESC` — the list order and not a relevance one, because nothing was
 * searched for. Nothing here re-sorts it.
 *
 * The ref goes through `notePath`, so it is one percent-encoded segment exactly as it is for every
 * other ref-taking call, and the suffix is appended outside the encoding.
 */
export async function listBacklinks(ref: string, options: Options = {}): Promise<Note[]> {
  const payload = await apiRequest<NoteList>(`${notePath(ref)}/backlinks`, options)
  return payload.notes
}

/**
 * Every wikilink this note's body currently contains, resolved as far as `/links` managed —
 * `GET /notes/{ref}/links` (KAN-566), read by KAN-567's editor pill.
 *
 * Unlike {@link listBacklinks}, this one is **not** the note noun: `/links` answers with `LinkRead`
 * rows, because a pill needs `target_kind`/`target_ref`/`resolved_ref`/`title`/`column` and a note
 * record has none of the first two. So `lib/types.ts` gained a `Link` interface for this call alone.
 *
 * May reach pandan for the pandan-shaped rows (`backend/app/api/links.py`) and never fails for that
 * reason — an unreachable pandan degrades every pandan-kind row to unresolved rather than refusing
 * the request (ADR 0003, Q26). A note-to-note edge never depends on pandan at all.
 */
export async function listLinks(ref: string, options: Options = {}): Promise<Link[]> {
  const payload = await apiRequest<LinkList>(`${notePath(ref)}/links`, options)
  return payload.links
}

/**
 * Every version of this note's body, newest first — `GET /notes/{ref}/versions` (R13/KAN-1064),
 * read by KAN-1065's History tab.
 *
 * **Full records, same as everything else this module returns.** The backend's design call
 * (`backend/app/api/schemas.py`'s `NoteVersionRead`) is that a version's whole `body` rides along
 * in the list response — a note body is small prose, and a preview is a selection over rows
 * already in hand, not a second request. `restoreVersion` below is what turns that selection into
 * a write.
 */
export async function listVersions(ref: string, options: Options = {}): Promise<NoteVersion[]> {
  const payload = await apiRequest<NoteVersionList>(`${notePath(ref)}/versions`, options)
  return payload.versions
}

/**
 * Restore a version: sugar over {@link updateNote}, and it must stay sugar (ADR 0008's `moveNote`
 * is the precedent — a restore is not a route, it is a `PATCH` whose `body` came from history).
 *
 * `ifUpdatedAt` is **required** here, unlike `updateNote`'s own optional precondition: a restore
 * is a write a person did not just type, aimed from a rail beside the editor rather than from
 * inside it, and BREADBOARD.md's R13 is explicit that it "goes through the same 409 precondition
 * as any other edit" — the caller (`HistoryPanel.svelte`) always has the open note's `updated_at`
 * in hand, so there is no honest reason for this one call site to spell the unguarded write.
 */
export function restoreVersion(
  ref: string,
  body: string,
  ifUpdatedAt: string,
  options: Options = {},
): Promise<Note> {
  return updateNote(ref, { body, if_updated_at: ifUpdatedAt }, options)
}

/** Delete the note. `204`, no body. The ref is never reused (ADR 0008). */
export async function deleteNote(ref: string, options: Options = {}): Promise<void> {
  await apiRequest<null>(notePath(ref), { ...options, method: 'DELETE' })
}
